### ISSUE TITLE
Button styles wrongly persist after toggling from light to dark, or dark to light

### ISSUE THREAD

**Issue by Nantris** [react-native-elements/react-native-elements#3665]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Explain what you did

Theming buttons with the theme provider and toggling the dark/light mode.

### Expected behavior

The colors should be changed to those defined under `theme.components`

### Describe the bug

Same bug as #3655, but for buttons

### Steps To Reproduce

```Markdown
Add a button with styles defined in the RneTheme, then toggle dark/light mode and note that the styles do not update
```

### Screenshots

Toggling from light mode to dark mode, or reverse, persists the originally loaded styles for background and text colors

## Correct button light mode

![image](https://user-images.githubusercontent.com/6835891/197883102-dfdf746f-dc11-4210-bed7-83d55850587a.png)

## Incorrect button dark mode (after loading app in light mode and switching to dark)
![image](https://user-images.githubusercontent.com/6835891/197883180-c707c0eb-8b98-4c80-ba62-14f7945da41f.png)

----

## Correct button dark mode

![image](https://user-images.githubusercontent.com/6835891/197883329-c54a39a1-4988-4980-8e37-98c808faf02a.png)

## Incorrect button light mode (after loading app in dark mode and switching to light)

![image](https://user-images.githubusercontent.com/6835891/197883447-3d4a7961-cf47-4018-86df-9bd0f5172e0b.png)

### Your Environment

<details>
<summary>`npx @rneui/envinfo`</summary>

      ```
        System:
    OS: Linux 5.15 Ubuntu 22.04.1 LTS 22.04.1 LTS (Jammy Jellyfish)
    CPU: (4) x64 AMD Ryzen 7 5800X 8-Core Processor
    Memory: 2.08 GB / 19.51 GB
    Container: Yes
    Shell: 5.8.1 - /usr/bin/zsh
  Binaries:
    Node: 16.17.0 - ~/.nvm/versions/node/v16.17.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.17.0/bin/yarn
    npm: 8.15.0 - ~/.nvm/versions/node/v16.17.0/bin/npm
    Watchman: 4.9.0 - /usr/bin/watchman
  Managers:
    Apt: 2.4.7 - /usr/bin/apt
    pip3: 22.2.2 - ~/.local/bin/pip3
  Utilities:
    CMake: 3.22.1 - /usr/bin/cmake
    Make: 4.3 - /usr/bin/make
    GCC: 11.2.0 - /usr/bin/gcc
    Git: 2.38.1 - /usr/bin/git
  SDKs:
    Android SDK:
      API Levels: 31, 32, 33
      Build Tools: 30.0.3, 31.0.0, 33.0.0
      System Images: android-31 | Google APIs Intel x86 Atom_64, android-31 | Google Play Intel x86 Atom_64
  IDEs:
    Nano: 6.2 - /usr/bin/nano
    VSCode: 1.68.1 - /usr/bin/code
  Languages:
    Bash: 5.1.16 - /usr/bin/bash
    Perl: 5.34.0 - /usr/bin/perl
    Python3: 3.10.6 - /usr/bin/python3
  Databases:
    SQLite: 3.32.2 - /home/me/Android/Sdk/platform-tools/sqlite3
  Browsers:
    Chrome: 106.0.5249.61
    Firefox: 106.0
  Monorepos:
    Yarn Workspaces: 1.22.19
    Lerna: 5.6.2
      ```

</details>

**Comment by Nantris** [comment]
@arpitBhalla any thoughts on this? I find this to be a higher priority issue as an app might use the `ListItem`s only in one or two places, but buttons are rather ubiquitous.

**Comment by arpitBhalla** [comment]
Would it be possible for you to share a [Snack](https://snack.expo.dev/@arpitbhalla/react-native-elements)?

**Comment by Nantris** [comment]
Whoops Github never told me about this message. Thanks for your reply @arpitBhalla.

[Here's a snack showing the issue](https://snack.expo.dev/@recollectr/react-native-elements?platform=android)


You can see the intended color in the console - but it never changes in the UI.

I tried setting a `key` on the `ThemeProvider` - but that leads to the entire component tree beneath being remounted - which is certainly undesirable.

![image](https://user-images.githubusercontent.com/6835891/202874237-3c9e3b04-1479-4c28-85b1-b9258a27c531.png)

**Comment by arpitBhalla** [comment]
Hey, I have updated the [snack](https://snack.expo.dev/@arpitbhalla/3665?platform=android)

would recommend using custom colors for such cases.

**Comment by Nantris** [comment]
Thanks for the continued work @arpitBhalla. That's workable, but it's really clunky. Why does it need to be set there for it to take effect? It's going to lead to a lot of duplicate code for every property that needs to be addressed like that.

**Edit:** Removed incorrect issue report/snack

**Comment by Nantris** [comment]
I don't mean to be rude at all in saying this - this workaround is awful. It makes simple code nightmarish to maintain. **And most importantly, it does not seem to actually work in our app.**

 I confirmed the colors are defined as expected, but in the `Button: (props, { colors }) => themeButton(motif, colors),` code, the background code is always the light mode, not the dark mode. When light/dark is toggled the buttons stay the same.

_I hope there will be another solution long term?_

Simple code like this:

```js
const motif = theme.activeMode();
return createTheme({
  Button: themeButton(motif)
})

const themeButton = (motif) => ({ ... })
```

Becomes a mess like this:

```js
const { lightMode, darkMode } = theme;
return createTheme({
    lightColors: {
      buttonBackground: lightMode.button.background,
    },
    darkColors: {
      buttonBackground: darkMode.button.background,
    }
  components: {
    Button: (props, { colors }) => themeButton(motif, colors),
  }
})

// Function now has to manage two sets of colors, even though the colors really are just duplicates
const themeButton = (motif, colors) => ({ ... })
```

It's actually a substantially larger mess than that code conveys, and dramatically worse on top of that since we need to do this for no fewer than three components, and any others that surprise us down the road (`Button`, `Icon`, `ListItem`.) Then we also have to do it for any text colors in addition to the background colors, and then what else will come up down the road? It makes relying on this library seem very risky.

We've been using React Native Elements for years but if these theming issues don't receive a better resolution then we'll have to look into changing libraries.

**Comment by Nantris** [comment]
Is there a way I can help with this?

**Comment by arpitBhalla** [comment]
I agree, this doesn't look clean and I never liked this approach myself, but I thought it will keep things simple. I really want to make it better.

Alternatively, this can be done with current version, but it's still duplicating the code.
```tsx
const themeSwitch =
  <T, P>({ $light, $dark }: { $light?: T; $dark?: T }) =>
  (props, theme) => {
    if (theme.mode === 'dark') {
      return $dark;
    }
    return $light;
  };
```
And use it like in this way (see [snack](https://snack.expo.dev/@arpitbhalla/3665-i?platform=android))
```tsx
export const theme = createTheme({
  components: {
    Button: themeSwitch({
      $dark: {
        buttonStyle: {
          backgroundColor: 'red',
        },
      },
      $light: {
        buttonStyle: {
          backgroundColor: 'blue',
        },
      },
    }),
	Text: themeSwitch({$light:{},$dark:{}})
  },
});
```
We can have nested colors as well, but avoid overwriting predefined color keys
```tsx
declare module '@rneui/themed' {
  interface Colors {
    button: {
      background: string;
      color: string;
    };
  }
}
```
And can do (similar to your snippet) (see [snack](https://snack.expo.dev/@arpitbhalla/3665-ii))
```tsx
const { lightMode, darkMode } = themeModes;
const theme = createTheme({
  lightColors: {
    button: lightMode.button,
  },
  darkColors: {
    button: darkMode.button,
  },
  components: {
    Button: (props, { colors }) => ({
      buttonStyle: {
        backgroundColor: colors.button.background,
      },
      titleStyle: {
        color: colors.button.color,
      },
    }),
  },
});
```

**Comment by arpitBhalla** [comment]
ern

**Comment by arpitBhalla** [comment]
Can also do in this way, (not yet implemented)

```tsx
export const theme = createTheme({
  components: {
    Button: {
      buttonStyle: {
		// 
        backgroundColor: 'colors.button.background',
      },
    },
  },
});
```


PS. accidentally pressed some key & closed it 🫠

**Comment by Nantris** [comment]
Thanks for your reply and your hard work @arpitBhalla!

In your last example, did you mean for this to be a string? `'colors.button.background'`

All of the other possibilities above duplicate code quite extensively.

If we could do something like below (not relying on setting RNE theme colors), that would be ideal:

```js
export const theme = createTheme({
  components: {
    Button: {
      buttonStyle: {
        backgroundColor: ourTheme.activeMode().button.background
      },
    },
  },
});
```

Needing to duplicate our color keys into RNEUI's theme is the main deal-breaker in my view, because it's a task without any real boundaries - the more styles we need, to more duplicated work we'd have to do.

I don't mind leaving these issues unfixed for up to a couple months in our app since it's still in alpha - I just need to know that a fix is planned.


----

What's the current limitation that prevents [my originally shared code](https://snack.expo.dev/@recollectr/react-native-elements?platform=android) from updating the components when the theme mode changes between light/dark? And the same issue affects the button's text `color`. Couldn't React's `Context` be used to get these elements to update in response to theming changes?

I understand there may be some performance limitations/optimizations at play here; I'd understand if we couldn't dynamically update theme colors after setting them because deep object equality comparisons are expensive - but I'd definitely expect light vs dark colors to be toggled (and especially since every other component I've tried besides `Button`, `Icon`, and `ListItem` already work as expected.)

What's the common thread between these elements that's causing them not to update in response to dark/light toggling? Is there anything I can do to help with the issue?

Thanks again very much @arpitBhalla

**Comment by arpitBhalla** [comment]
> In your last example, did you mean for this to be a string? 'colors.button.background'

Yes, but we still need to specify `lightColors` and `darkColors`

In the snippet you provided, how `ourTheme` (should) look like, and how `activeMode()` (should) work?

---

ThemeProvider isn't listening/updating `theme` prop change, will make changes for this. Still it can be done via `replaceTheme` .

```tsx
const useChangeTheme=(color:string)=>{
  const {replaceTheme}=useTheme()

  useEffect(()=>{
    replaceTheme(createCustomTheme(color))
  },[color])
}
```


See [snack](https://snack.expo.dev/@arpitbhalla/3665-iii) (might be weird way of doing it 🥲, and cause multiple renders)

I understand having so many colors in context will make it large and you might not use them in components(?). I am open for any suggestions. @Slapbox

**Comment by Nantris** [comment]
@arpitBhalla thanks again for all your hard work! I tried #3732 via patch-package and it works as expected for us! Hooray!

If it doesn't have any big performance downsides (and I don't think it should) - then I'm definitely in favor of getting that merged!

I think it's a much better approach, because a more mature application is likely to have its own internal theme/palette definitions that shouldn't require duplication into the RNEUI theme. Your change in #3732 lets us use our pre-existing color palette to style elements in the RNE theme object and it "just works." 

It now matches how all the other `ThemeProvider` components I've used function.

**Comment by arpitBhalla** [comment]
>  it "just works."

Feel free to suggest how it can be improved.

**Comment by Nantris** [comment]
@arpitBhalla I think your PR looks good as it is! That simple change you've suggested will fix our problem once merged.
